### PR TITLE
FOUNDATIONS: Document RetrieveAccessPolicyByName 

### PR DESCRIPTION
**Pull request by MarcusEnrique**
closes [AB#17975](https://dev.azure.com/NELAnalytics/aef6c175-110f-4ba4-82fc-70761f459236/_workitems/edit/17975)
closes [AB#18078](https://dev.azure.com/NELAnalytics/aef6c175-110f-4ba4-82fc-70761f459236/_workitems/edit/18078)
closes [AB#18073](https://dev.azure.com/NELAnalytics/aef6c175-110f-4ba4-82fc-70761f459236/_workitems/edit/18073)

### CODE REVIEW

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs [hunk 1]
```diff
@@ -57,6 +58,22 @@ public ValueTask<List<string>> RetrieveAllAccessPoliciesFromContainerAsync(strin
...
+            ValidateStorageArgumentsOnRetrieveAccessPolicies(container);
+            
+            return await this.blobStorageBroker.RetrieveAllAccessPoliciesAsync(container);
+        });
+
+        public ValueTask<Policy> RetrieveAccessPolicyByNameAsync(string container, string policyName) =>
+        TryCatch(async () =>
+        {
+            ValidateStorageArgumentsOnRetrieveAccessPolicyByName(container, policyName);
+
```
Comment (MichaelOSullivanAnswer) [inline ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs:73]: You will need to add some validation here to ensure the name exists. Look at [https://github.com/NHSISL/ISL.Providers.Storages/blob/main/ISL.Providers.Storages.AzureBlobStorage/Services/Foundations/Storages/StorageService.cs](url) for example

File: ISL.ReIdentification.Core/Brokers/Storages/Blob/BlobStorageBroker.cs [hunk 2]
```diff
@@ -62,6 +62,9 @@ public async ValueTask<List<string>> RetrieveAllAccessPoliciesFromContainerAsync
         public async ValueTask<List<Policy>> RetrieveAllAccessPoliciesAsync(string container) =>
             throw new NotImplementedException();
 
+        public async ValueTask<Policy> RetrieveAccessPolicyByNameAsync(string container, string policyName) =>
+        throw new NotFiniteNumberException();
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core/Brokers/Storages/Blob/BlobStorageBroker.cs:66]: fix indentation

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.Validations.cs [hunk 3]
```diff
@@ -38,6 +38,13 @@ private static void ValidateStorageArgumentsOnRetrieveAccessPolicies(string cont
             Validate((Rule: IsInvalid(container), Parameter: "Container"));
         }
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.Validations.cs:39]: could be a one liner

```cs
        private static void ValidateStorageArgumentsOnRetrieveAccessPolicies(string container) =>
            Validate((Rule: IsInvalid(container), Parameter: "Container"));
```

File: ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.cs [hunk 4]
```diff
@@ -88,6 +88,21 @@ private static List<Policy> GetPolicies() =>
             },
         };
 
+        private static Policy GetPolicy()
+        {
+            return new Policy
+            {
+                PolicyName = "default",
+                Permissions = new List<string>
+        {
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core.Tests.Unit/Services/Foundations/Documents/DocumentsTests.cs:97]: fix indentation

File: ISL.ReIdentification.WebApps.Portals/ISL.ReIdentification.Portals.Server/appsettings.json [hunk 5]
```diff
@@ -63,7 +63,10 @@
     "Container": "projects",
     "LandingFolder": "outbox",
     "PickupFolder": "inbox",
-    "ErrorFolder": "error"
+    "ErrorFolder": "error",
+    "ServiceUri": "override_this_in_your_appsettings.Development.json_file_or_environment_variables",
+    "StorageAccountName": "override_this_in_your_appsettings.Development.json_file_or_environment_variables",
+    "StorageAccountAccessKey": "override_this_in_your_appsettings.Development.json_file_or_environment_variables"
```
Comment (cjdutoit) [inline ISL.ReIdentification.WebApps.Portals/ISL.ReIdentification.Portals.Server/appsettings.json:69]: Remove this section

File: ISL.ReIdentification.WebApps.Configurations/ISL.ReIdentification.Configurations.Server/appsettings.json [hunk 6]
```diff
@@ -56,7 +56,10 @@
     "Container": "projects",
     "LandingFolder": "outbox",
     "PickupFolder": "inbox",
-    "ErrorFolder": "error"
+    "ErrorFolder": "error",
+    "ServiceUri": "override_this_in_your_appsettings.Development.json_file_or_environment_variables",
```
Comment (cjdutoit) [inline ISL.ReIdentification.WebApps.Configurations/ISL.ReIdentification.Configurations.Server/appsettings.json:60]:
```suggestion
    "Container": "projects",
    "LandingFolder": "outbox",
    "PickupFolder": "inbox",
    "ErrorFolder": "error",
    "ServiceUri": "override_this_in_your_appsettings.Development.json_file_or_environment_variables",
    "StorageAccountName": "override_this_in_your_appsettings.Development.json_file_or_environment_variables",
    "StorageAccountAccessKey": "override_this_in_your_appsettings.Development.json_file_or_environment_variables"
  },
```

File: ISL.ReIdentification.WebApps.Configurations/ISL.ReIdentification.Configurations.Server/appsettings.json [hunk 7]
```diff
@@ -56,7 +56,10 @@
...
+    "StorageAccountName": "override_this_in_your_appsettings.Development.json_file_or_environment_variables",
+    "StorageAccountAccessKey": "override_this_in_your_appsettings.Development.json_file_or_environment_variables"
   },
   "ReIdentificationProviderOfflineMode": false,
```
Comment (cjdutoit) [inline ISL.ReIdentification.WebApps.Configurations/ISL.ReIdentification.Configurations.Server/appsettings.json:64]: Remove

File: ISL.ReIdentification.Infrastructure/Services/ScriptGenerationService.cs [hunk 8]
```diff
@@ -57,6 +57,9 @@ public void GenerateBuildScript(string branchName, string projectName, string do
                                 { "NOTIFICATIONCONFIGURATIONS__IMPERSONATIONAPPROVEDREQUESTTEMPLATEID", "${{ secrets.NOTIFICATIONCONFIGURATIONS__IMPERSONATIONAPPROVEDREQUESTTEMPLATEID }}" },
                                 { "NOTIFICATIONCONFIGURATIONS__IMPERSONATIONDENIEDREQUESTTEMPLATEID", "${{ secrets.NOTIFICATIONCONFIGURATIONS__IMPERSONATIONDENIEDREQUESTTEMPLATEID }}" },
                                 { "NOTIFICATIONCONFIGURATIONS__IMPERSONATIONPENDINGAPPROVALREQUESTTEMPLATEID", "${{ secrets.NOTIFICATIONCONFIGURATIONS__IMPERSONATIONPENDINGAPPROVALREQUESTTEMPLATEID }}" },
+                                { "AZUREBLOBSTORECONFIGURATIONS__STORAGEACCOUNTACCESSKEY", "${{ secrets.AZUREBLOBSTORECONFIGURATIONS__STORAGEACCOUNTACCESSKEY }}" },
+                                { "AZUREBLOBSTORECONFIGURATIONS__STORAGEACCOUNTNAME", "${{ secrets.AZUREBLOBSTORECONFIGURATIONS__STORAGEACCOUNTNAME }}"},
+                                { "AZUREBLOBSTORECONFIGURATIONS__SERVICEURI", "${{ secrets.AZUREBLOBSTORECONFIGURATIONS__SERVICEURI }}"},
```
Comment (cjdutoit) [inline ISL.ReIdentification.Infrastructure/Services/ScriptGenerationService.cs:62]:
```suggestion
                                { "PROJECTSTORAGECONFIGURATION__STORAGEACCOUNTACCESSKEY", "${{ secrets.PROJECTSTORAGECONFIGURATION__STORAGEACCOUNTACCESSKEY }}" },
                                { "PROJECTSTORAGECONFIGURATION__STORAGEACCOUNTNAME", "${{ secrets.PROJECTSTORAGECONFIGURATION__STORAGEACCOUNTNAME }}"},
                                { "PROJECTSTORAGECONFIGURATION__SERVICEURI", "${{ secrets.PROJECTSTORAGECONFIGURATION__SERVICEURI }}"},
```

File: ISL.ReIdentification.Functions/Program.cs [hunk 9]
```diff
@@ -15,15 +15,17 @@
 using ISL.Providers.ReIdentification.OfflineFileSources.Models;
 using ISL.Providers.ReIdentification.OfflineFileSources.Providers.OfflineFileSources;
 using ISL.Providers.Storages.Abstractions;
+using ISL.Providers.Storages.AzureBlobStorage.Models;
```
Comment (cjdutoit) [inline ISL.ReIdentification.Functions/Program.cs:18]: This using should be for the new model `AccessPolicy`

File: ISL.ReIdentification.WebApps.Portals/ISL.ReIdentification.Portals.Server/Program.cs [hunk 10]
```diff
@@ -242,12 +243,24 @@ private static void AddProviders(IServiceCollection services, IConfiguration con
...
+            AzureBlobStoreConfigurations azureBlobStoreConfigurations = new AzureBlobStoreConfigurations
+            {
+                ServiceUri = projectStorageConfiguration.ServiceUri,
+                StorageAccountName = projectStorageConfiguration.StorageAccountName,
+                StorageAccountAccessKey = projectStorageConfiguration.StorageAccountAccessKey
+            };
+
             services.AddSingleton(notificationConfigurations);
             services.AddSingleton(notifyConfigurations);
+            services.AddSingleton(azureBlobStoreConfigurations);
```
Comment (cjdutoit) [inline ISL.ReIdentification.WebApps.Portals/ISL.ReIdentification.Portals.Server/Program.cs:259]:
```suggestion
            services.AddSingleton(notificationConfigurations);
            services.AddSingleton(notifyConfigurations);

            ProjectStorageConfiguration projectStorageConfiguration = configuration
                .GetSection("projectStorageConfiguration")
                    .Get<ProjectStorageConfiguration>();

            AzureBlobStoreConfigurations projetcsBlobStoreConfigurations = new AzureBlobStoreConfigurations
            {
                ServiceUri = projectStorageConfiguration.ServiceUri,
                StorageAccountName = projectStorageConfiguration.StorageAccountName,
                StorageAccountAccessKey = projectStorageConfiguration.StorageAccountAccessKey
            };

            services.AddSingleton(projetcsBlobStoreConfigurations);
```

File: ISL.ReIdentification.Functions/Program.cs [hunk 11]
```diff
@@ -91,20 +93,32 @@ private static async Task Main(string[] args)
...
+        AzureBlobStoreConfigurations azureBlobStoreConfigurations = new AzureBlobStoreConfigurations
+        {
+            ServiceUri = projectStorageConfiguration.ServiceUri,
+            StorageAccountName = projectStorageConfiguration.StorageAccountName,
+            StorageAccountAccessKey = projectStorageConfiguration.StorageAccountAccessKey
+        };
+
         services.AddSingleton(notificationConfigurations);
         services.AddSingleton(notifyConfigurations);
+        services.AddSingleton(azureBlobStoreConfigurations);
```
Comment (cjdutoit) [inline ISL.ReIdentification.Functions/Program.cs:117]:
```suggestion
            services.AddSingleton(notificationConfigurations);
            services.AddSingleton(notifyConfigurations);

            ProjectStorageConfiguration projectStorageConfiguration = configuration
                .GetSection("projectStorageConfiguration")
                    .Get<ProjectStorageConfiguration>();

            AzureBlobStoreConfigurations projetcsBlobStoreConfigurations = new AzureBlobStoreConfigurations
            {
                ServiceUri = projectStorageConfiguration.ServiceUri,
                StorageAccountName = projectStorageConfiguration.StorageAccountName,
                StorageAccountAccessKey = projectStorageConfiguration.StorageAccountAccessKey
            };

            services.AddSingleton(projetcsBlobStoreConfigurations);
```

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs [hunk 12]
```diff
@@ -47,38 +46,51 @@ public ValueTask RemoveDocumentByFileNameAsync(string fileName, string container
...
         public ValueTask<List<Policy>> RetrieveAllAccessPoliciesAsync(string container) =>
         TryCatch(async () =>
         {
             ValidateStorageArgumentsOnRetrieveAccessPolicies(container);
-            
+
             return await this.blobStorageBroker.RetrieveAllAccessPoliciesAsync(container);
         });
 
+        public ValueTask<Policy> RetrieveAccessPolicyByNameAsync(string container, string policyName) =>
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs:68]:
```suggestion
        public ValueTask<AccessPolicy> RetrieveAccessPolicyByNameAsync(string container, string policyName) =>
```

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs [hunk 13]
```diff
@@ -47,38 +46,51 @@ public ValueTask RemoveDocumentByFileNameAsync(string fileName, string container
...
+        TryCatch(async () =>
+        {
+            ValidateStorageArgumentsOnRetrieveAccessPolicyByName(container, policyName);
+
+            List<string> policyNames = await this.blobStorageBroker
+                .RetrieveListOfAllAccessPoliciesAsync(container);
+
+            ValidateAccessPolicyExists(policyName, policyNames);
+
+            return await this.blobStorageBroker.RetrieveAccessPolicyByNameAsync(container, policyName);
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs:78]: Add mapping logic to convert from `Policy` to `AccessPolicy`

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs [hunk 14]
```diff
@@ -47,38 +46,51 @@ public ValueTask RemoveDocumentByFileNameAsync(string fileName, string container
...
+            return await this.blobStorageBroker.RetrieveListOfAllAccessPoliciesAsync(container);
         });
 
         public ValueTask<List<Policy>> RetrieveAllAccessPoliciesAsync(string container) =>
         TryCatch(async () =>
         {
             ValidateStorageArgumentsOnRetrieveAccessPolicies(container);
-            
+
             return await this.blobStorageBroker.RetrieveAllAccessPoliciesAsync(container);
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs:65]: Add mapping logic to convert from `Policy` to `AccessPolicy`

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs [hunk 15]
```diff
@@ -47,38 +46,51 @@ public ValueTask RemoveDocumentByFileNameAsync(string fileName, string container
             await this.blobStorageBroker.DeleteFileAsync(fileName, container);
         });
 
-        public ValueTask<string> GetDownloadLinkAsync(string fileName, string container) =>
-           throw new NotImplementedException();
+        public ValueTask CreateAndAssignAccessPoliciesAsync(string container, List<Policy> policies) =>
```
Comment (cjdutoit) [inline ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs:49]:
```suggestion
        public ValueTask CreateAndAssignAccessPoliciesAsync(string container, List<AccessPolicy> policies) =>
```

File: ISL.ReIdentification.WebApps.Portals/ISL.ReIdentification.Portals.Server/appsettings.json (hunk 5 above)
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.WebApps.Portals/ISL.ReIdentification.Portals.Server/appsettings.json:69]: Updated

File: ISL.ReIdentification.WebApps.Configurations/ISL.ReIdentification.Configurations.Server/Program.cs [hunk 16]
```diff
@@ -237,12 +238,24 @@ private static void AddProviders(IServiceCollection services, IConfiguration con
                 ApiKey = notificationConfigurations.ApiKey
             };
 
+            ProjectStorageConfiguration projectStorageConfiguration = configuration
+            .GetSection("projectStorageConfiguration")
+                .Get<ProjectStorageConfiguration>();
+
+            AzureBlobStoreConfigurations azureBlobStoreConfigurations = new AzureBlobStoreConfigurations
```
[the remark this answers is not in the document]
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.WebApps.Configurations/ISL.ReIdentification.Configurations.Server/Program.cs]: Updated

File: ISL.ReIdentification.Infrastructure/Services/ScriptGenerationService.cs (hunk 8 above)
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.Infrastructure/Services/ScriptGenerationService.cs]: Updated

File: ISL.ReIdentification.WebApps.Portals/ISL.ReIdentification.Portals.Server/Program.cs (hunk 10 above)
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.WebApps.Portals/ISL.ReIdentification.Portals.Server/Program.cs]: Updated

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs (hunk 12 above)
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs]: Updated

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs (hunk 13 above)
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs]: Updated

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs [hunk 17]
```diff
@@ -47,38 +46,51 @@ public ValueTask RemoveDocumentByFileNameAsync(string fileName, string container
...
+        public ValueTask<List<string>> RetrieveListOfAllAccessPoliciesAsync(string container) =>
         TryCatch(async () =>
         {
             ValidateStorageArgumentsOnRetrieveAccessPolicies(container);
 
-            return await this.blobStorageBroker.RetrieveAllAccessPoliciesFromContainerAsync(container);
+            return await this.blobStorageBroker.RetrieveListOfAllAccessPoliciesAsync(container);
         });
 
         public ValueTask<List<Policy>> RetrieveAllAccessPoliciesAsync(string container) =>
```
[the remark this answers is not in the document]
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs]: Updated

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs (hunk 14 above)
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs]: Updated

File: ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs (hunk 15 above)
Comment (MichaelOSullivanAnswer) [reply ISL.ReIdentification.Core/Services/Foundations/Documents/DocumentService.cs]: Updated

**Review by cjdutoit — APPROVED** [review]

**Pull request merged**